app/src/main/java/ru/gorshenev/themesstyles/presentation/ui/chat/middleware/GetQueueReactionMiddleware.kt

```diff
@@ -0,0 +1,75 @@
+package ru.gorshenev.themesstyles.presentation.ui.chat.middleware
+
+import io.reactivex.Observable
+import ru.gorshenev.themesstyles.data.repositories.chat.ChatMapper.toDomain
+import ru.gorshenev.themesstyles.data.repositories.chat.ChatMapper.toUi
+import ru.gorshenev.themesstyles.data.repositories.chat.ChatRepository
+import ru.gorshenev.themesstyles.presentation.base.recycler_view.ViewTyped
+import ru.gorshenev.themesstyles.presentation.mvi_core.Middleware
+import ru.gorshenev.themesstyles.presentation.ui.chat.ChatAction
+import ru.gorshenev.themesstyles.presentation.ui.chat.ChatInternalAction
+import ru.gorshenev.themesstyles.presentation.ui.chat.ChatState
+import ru.gorshenev.themesstyles.presentation.ui.chat.items.MessageLeftUi
+import ru.gorshenev.themesstyles.presentation.ui.chat.items.MessageRightUi
+
+class GetQueueReactionMiddleware(private val repository: ChatRepository) :
+    Middleware<ChatAction, ChatState> {
+
+    private var lastId = -1
+
+    override fun bind(
+        actions: Observable<ChatAction>,
+        state: Observable<ChatState>
+    ): Observable<ChatAction> {
+        return actions.ofType(ChatAction.GetQueueReaction::class.java)
+            .withLatestFrom(state) { action, currentState -> action to currentState }
+            .switchMap { (action, state) ->
+                repository.getQueueReactions(action.queueId, action.lastId)
+                    .toObservable()
+                    .retry()
+                    .flatMap { response ->
+                        val event = response.events.single()
+                        lastId = event.id
+                        repository.getMessage(event.messageId).toObservable()
+                    }
+                    .flatMap { response ->
+                        Observable.just(response.message)
+                            .flatMapCompletable { repository.saveMessage(it, action.topicName) }
```

Comment:
repository.getQueueReactions(action.queueId, action.lastId)
                    .retry()
                    .flatMapObservable { response ->
                        val event = response.events.single()
                        lastId = event.id
                        repository.getMessage(event.messageId).toObservable()
                    }
                    .withLatestFrom(state) { response, currentState -> response to currentState }
                    .filter { (_, currentState) -> currentState is ChatState.Result }
                    .flatMap<ChatAction> { (response, currentState) ->
                        val updatedItems = updateMessage(
                            newMessage = listOf(response.message).toDomain().toUi().first(),
                            currentItems = (currentState as ChatState.Result).items
                        )
                        val resultAction = ChatAction.GetQueueReaction(action.queueId, lastId, action.topicName, updatedItems)
                        repository.saveMessage(response.message, action.topicName)
                            .toSingleDefault(resultAction)
                            .toObservable()
                    }
                    .onErrorReturn { ChatInternalAction.LoadError(it) }

Reply:
+